src/plugins/discover/public/application/components/source_viewer/source_viewer.tsx

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+import React, { useEffect, useState } from 'react';
+import { monaco } from '@kbn/monaco';
+import { EuiEmptyPrompt, EuiLoadingSpinner } from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { useEsDocSearch } from '../doc/use_es_doc_search';
+import { DocProps } from '../doc/doc';
+import { JsonCodeEditorCommon } from '../json_code_editor/json_code_editor_common';
+import { ElasticRequestState } from '../doc/elastic_request_state';
+
+interface SourceViewerProps {
+  docProps: DocProps;
+  hasLineNumbers: boolean;
+  width?: number;
+}
+
+export const SourceViewer = ({ docProps, width, hasLineNumbers }: SourceViewerProps) => {
+  const [editor, setEditor] = useState<monaco.editor.IStandaloneCodeEditor>();
+  const [jsonValue, setJsonValue] = useState<string>('');
+
+  const [reqState, hit] = useEsDocSearch({ ...docProps });
+
+  useEffect(() => {
+    if (reqState === ElasticRequestState.Found) {
+      setJsonValue(JSON.stringify(hit, null, 2));
+    }
+  }, [reqState, hit]);
+
+  // setting editor height based on lines height and count to stretch and fit its content
+  useEffect(() => {
+    if (!editor) {
+      return;
+    }
+    const editorElement = editor.getDomNode();
+
+    if (!editorElement) {
+      return;
+    }
+
+    const lineHeight = editor.getOption(monaco.editor.EditorOption.lineHeight);
+    const lineCount = editor.getModel()?.getLineCount() || 1;
+    const height = editor.getTopForLineNumber(lineCount + 1) + lineHeight;
+    editorElement.style.height = `${height}px`;
+    editor.layout();
+  }, [editor, jsonValue]);
+
+  const loadingState = <EuiLoadingSpinner size="m" />;
+
+  const errorMessageTitle = (
+    <h2>
+      {i18n.translate('discover.sourceViewer.errorMessageTitle', {
+        defaultMessage: 'An Error Occurred',
+      })}
+    </h2>
+  );
+  const errorMessage = (
+    <p>
+      {i18n.translate('discover.sourceViewer.errorMessage', {
+        defaultMessage: 'Could not fetch source at this time. Refresh the tab to try again.',
+      })}
+    </p>
+  );
+  const errorState = (
+    <EuiEmptyPrompt iconType="alert" title={errorMessageTitle} body={errorMessage} />
+  );
+
+  if (reqState === ElasticRequestState.Loading) {
+    return loadingState;
+  }
+
```

Comment:
So hopefully most users won't see this loading spinner, but if network is slow, there are 2 things we should have a closer at:

https://user-images.githubusercontent.com/463851/121925175-ef36a600-cd3c-11eb-98e2-629abd4234c4.mp4

* There is a short glimpse of `null` when the JSON view is rendered.
* @andreadelrio could you have a look at the design of the loading spinner, maybe it should display some text, to indicate what's going on? Just a note, it needs work for both the inline display in Classic and the flyout version of New table

Reply:
@majagrubic Suggesting this for the loading bit
<img width="745" alt="Frame 74" src="https://user-images.githubusercontent.com/4016496/122443192-2fbc3c80-cf54-11eb-82c2-d461999edac0.png">

You can get this done by reusing most of the code/styles for `.euiDataGrid__loading`. I was making the changes myself but for some reason the styles weren't being applied.